### ISSUE TITLE
Double request when using basic auth

### ISSUE THREAD

**Issue by akitoeki** [daltoniam/SwiftHTTP#150]
Double requests: first one with no credentials that always fail, then send the one with credential after that, resulting in two requests

**Comment by daltoniam** [comment]
Interesting. Can you provide an example?

**Comment by akitoeki** [comment]
Well, it's just as simple as the demo code on the readme: 

```swift
func checkInLocation(location: CLLocation, completed: (()->Void)?) {
        var endpoint = SERVER_ADDRESS+"/drivers/checkin"
        let request = HTTPTask()
        
        var attempted = false
        let email = self.email
        let password = self.password
        
        request.auth = {(challenge: NSURLAuthenticationChallenge) in
            if (!attempted) {
                attempted = true
                return NSURLCredential(user: email, password: password, persistence: .None)
            }
            return nil
        }
        
        
        var params: Dictionary<String, AnyObject> = ["latitude":location.coordinate.latitude, "longtitude":location.coordinate.longitude]
        println("Requesting to \(endpoint) with params: \n \(params)")
        
        request.requestSerializer = JSONRequestSerializer()
        request.responseSerializer = JSONResponseSerializer()
        
        request.POST(endpoint, parameters: params) { (res) -> Void in
            if (res.error == nil) {
                dispatch_async(dispatch_get_main_queue(), { () -> Void in
                    completed?()
                })
            }
        }
    }
```

ExpressJS log: 
```
POST /drivers/checkin 401 2.703 ms - -
POST /drivers/checkin 200 5.013 ms - 95
```

Always 401 (Auth error) before 200 :dancer:

**Comment by daltoniam** [comment]
I think this is the standard behavior for the NSURLSession and the challenge delegate. It doesn't know that basic auth is needed until after the request fails. I looked closer at the docs and came across:

https://developer.apple.com/library/mac/documentation/Foundation/Reference/NSURLSessionConfiguration_class/#//apple_ref/occ/instp/NSURLSessionConfiguration/URLCredentialStorage

It might make sense to have credentials for the auth block added to this shared storage. The default NSURLSessionConfiguration also uses a shared object according to:

https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Classes/NSURLCredentialStorage_Class/#//apple_ref/occ/instm/NSURLCredentialStorage

It might be worth a test of adding the `NSURLCredential ` of the email and password to the shared storage and see if that fixes the double requests. If that does we might just add something to the docs about how that works and think of a nicer way to handle this in the Swift 2 branch (since that will be coming out shortly).

**Comment by akitoeki** [comment]
Thanks, i'll try looking in that direction too, and since I'm still stuck with 1.2 for my project, might as well do a pull request on your 1.2 version.

Cheers

**Comment by fform** [comment]
@daltoniam I know when I'm using basic auth, it's known beforehand and there will be no need for a double request. Especially for when multiple endpoints from the same domain will require basic auth.

I'm currently dealing with this by generating the base64 string before each request like
```swift
let authData = "Username:MySecretPassword!".dataUsingEncoding(NSASCIIStringEncoding, allowLossyConversion: false)
let authValue = "Basic \(authData!.base64EncodedStringWithOptions(NSDataBase64EncodingOptions.Encoding76CharacterLineLength))"
```
then I make a mutable request and add a custom header
```swift
req.addValue(authValue, forHTTPHeaderField: "Authorization")
```

This is for the swift 2.0 version. It was actually a little easier before since you were always making a mutable request but now I'm making a mutable request and making an Operation out of it with something like `HTTP(req)`. I have a helper function making this request and adding the auth header, but there must be a better way, right?

**Comment by daltoniam** [comment]
@fform Thanks for sharing. I had looked at doing but according to the NSMutableRequest docs it says not to override a few specific headers and `Authorization` was one of them. I don't think it will cause an issue in most cases, but just trying to obey the docs (if possible :smile: ).

https://developer.apple.com/library/mac/documentation/Cocoa/Reference/Foundation/Classes/NSMutableURLRequest_Class/#//apple_ref/occ/instm/NSMutableURLRequest/addValue:forHTTPHeaderField:

Yeah for the Swift 2.0 version I tried to integrate with the built in APIs instead of completely replacing them as I did in version 1. I feel this saved a lot of code and made it easier to extend. I do agree it was less code to change/update headers in v1 since it completely shielded from interacting with `NSMutableRequest`. I guess we could should an optional `headers` argument to the `HTTP` factory methods to make adding custom header easier.
